core/src/test/java/com/linecorp/armeria/client/RedirectingClientTest.java

```diff
@@ -0,0 +1,287 @@
+
+            sb.service("/loop", (ctx, req) -> HttpResponse.ofRedirect("loop1"))
+              .service("/loop1", (ctx, req) -> HttpResponse.ofRedirect("loop2"))
+              .service("/loop2", (ctx, req) -> HttpResponse.ofRedirect("loop"));
+
+            sb.service("/differentHttpMethod", (ctx, req) -> {
+                if (ctx.method() == HttpMethod.GET) {
+                    return HttpResponse.of(200);
+                } else {
+                    assertThat(ctx.method()).isSameAs(HttpMethod.POST);
+                    return HttpResponse.ofRedirect(HttpStatus.SEE_OTHER, "/differentHttpMethod");
+                }
+            });
+        }
+
+        private int otherHttpPort(ServiceRequestContext ctx) {
+            final Optional<ServerPort> serverPort =
+                    server.server().activePorts().values()
+                          .stream().filter(ServerPort::hasHttp)
+                          .filter(port -> port.localAddress().getPort() !=
+                                          ((InetSocketAddress) ctx.localAddress()).getPort())
+                          .findFirst();
+            assert serverPort.isPresent();
+            return serverPort.get().localAddress().getPort();
+        }
+    };
+
+    @BeforeEach
+    void setUp() {
+        requestCounter.set(0);
+    }
+
+    @Test
+    void redirect_successWithRetry() {
+        final AggregatedHttpResponse res = sendRequest(2);
+        assertThat(res.status()).isSameAs(HttpStatus.OK);
+    }
+
+    @Test
+    void redirect_failExceedingTotalAttempts() {
+        assertThatThrownBy(() -> sendRequest(1))
+                .hasCauseInstanceOf(TooManyRedirectsException.class)
+                .hasMessageContainingAll("maxRedirects: 1", "The original URI: ",
+                                         "Redirect URIs:", "fooRedirect1", "fooRedirect2");
+    }
+
+    private static AggregatedHttpResponse sendRequest(int maxRedirects) {
+        final WebClient client = WebClient.builder(server.httpUri())
+                                          .followRedirects(RedirectConfig.builder()
+                                                                         .maxRedirects(maxRedirects)
+                                                                         .build())
+                                          .build();
+        return client.get("/foo").aggregate().join();
+    }
+
+    @Test
+    void protocolChange() {
+        final WebClient client = WebClient.builder(server.httpUri())
+                                          .factory(ClientFactory.insecure())
+                                          .followRedirects()
+                                          .build();
+        assertThatThrownBy(() -> client.get("/protocolChange").aggregate().join())
+                .hasCauseInstanceOf(UnexpectedProtocolRedirectException.class)
+                .hasMessageContaining("redirectProtocol: https (expected: [http])");
```

Comment:
Is it better to allow changing protocols from HTTP to HTTPS if `allowProtocols()` is not specified?

Reply:
That's a good suggestion. Fixed. 😉